Source/OpalGraphics/CGPath.m

```diff
@@ -159,148 +193,395 @@ CGPoint CGPathGetCurrentPoint(CGPathRef path)
+
+  scale = fabs(x0) + fabs(x1) + fabs(y0) + fabs(y1);
+  if (!scale)
+    return 0;
+
+  scale /= 40000000.0;
+
+  /* Deal with the degenerate case to == from.  */
+  if (to.x == from.x && to.y == from.y)
+    {
+      if (x1 - x0 < scale && y1 - y0 < scale)
+	return winding_line(from, to, p);
+    }
+  else
+    {
+      double dx, dy;
+      double nx, ny;
+      double d0, d1, d2, d3;
+
+      /* Get the direction vector and the normal vector.  */
+      dx = to.x - from.x;
+      dy = to.y - from.y;
+      d0 = sqrt(dx * dx + dy * dy);
+      dx /= d0;
+      dy /= d0;
+      nx = dy;
+      ny = -dx;
+
+      /* Check that the distances along the direction vector are
+	 monotone.  */
+
+      d0 = from.x * dx + from.y * dy;
+      d1 = c1.x * dx + c1.y * dy;
+      d2 = c2.x * dx + c2.y * dy;
+      d3 = to.x * dx + to.y * dy;
+
+      if ((d3 > d2 && d2 > d1 && d1 > d0)
+	  || (d3 < d2 && d2 < d1 && d1 < d0))
+	{
+	  /* Check that the control points are close to the straigt line
+	     between from and to.  */
+	  d0 = to.x * nx + to.y * ny;
+	  d1 = c1.x * nx + c1.y * ny;
+	  d2 = c2.x * nx + c2.y * ny;
+
+	  if (fabs(d0 - d1) < scale && fabs(d0 - d2) < scale)
+	    {
+	      /* It's flat enough.  */
+	      return winding_line(from, to, p);
+	    }
+	}
+    }
+
+  {
+    /* Subdivide.  */
+    double_point m, l1, l2, r1, r2;
+
+    m.x = (from.x + to.x + 3 * (c1.x + c2.x)) / 8;
+    m.y = (from.y + to.y + 3 * (c1.y + c2.y)) / 8;
+
+    l1.x = (from.x + c1.x) / 2;
+    l1.y = (from.y + c1.y) / 2;
+
+    l2.x = (from.x + 2 * c1.x + c2.x) / 4;
+    l2.y = (from.y + 2 * c1.y + c2.y) / 4;
+
+    r2.x = (to.x + c2.x) / 2;
+    r2.y = (to.y + c2.y) / 2;
+
+    r1.x = (to.x + 2 * c2.x + c1.x) / 4;
+    r1.y = (to.y + 2 * c2.y + c1.y) / 4;
+
+    return winding_curve(from, m, l1, l2, p, depth + 1)
+           + winding_curve(m, to, r1, r2, p, depth + 1);
+  }
+}
+
+static int _OPPathGetWindingCount(CGPathRef path, CGPoint point)
+{
+  int total;
+  CGPathElementType type;
+  NSInteger count;
+  BOOL first;
+  CGPoint pts[3];
+  CGPoint first_p, last_p;
+  NSInteger i;
+
+  total = 0;
+  count = [path count];
+
+  if (count == 0)
+    return 0;
+
+  type = [path elementTypeAtIndex: 0 points: pts];
+  if (type != kCGPathElementMoveToPoint)
+    {
+      return 0;
+    }
+  last_p = first_p = pts[0];
+  first = NO;
+
+#define D(a) (double_point){a.x,a.y}
+  for (i = 1; i < count; i++)
+    {
+      type = [path elementTypeAtIndex: i points: pts];
+      switch(type)
+	{
+	  case kCGPathElementMoveToPoint:
+	    if (!first)
+	      {
+		total += winding_line(D(last_p), D(first_p), D(point));
+	      }
+	    last_p = first_p = pts[0];
+	    first = NO;
+	    break;
+	  case kCGPathElementAddLineToPoint:
+	    if (first)
+	      {
+		return 0;
+	      }
+	    total += winding_line(D(last_p), D(pts[0]), D(point));
+	    last_p = pts[0];
+	    break;
+	  case kCGPathElementAddCurveToPoint:

```

Comment:
I'm not sure on how to implement the winding algorithm for quadratic curves; for now I left it as a `FIXME`. Could you point me in any direction?